File: connectors/scripts/registry_generator.py
#!/usr/bin/env python3
"""
连接器注册表生成脚本
用于GitHub Actions和本地开发
"""

import json
import glob
import os
import sys
import argparse
import hashlib
from datetime import datetime, timezone
from pathlib import Path


def generate_registry(connectors_dir: str = ".", output_file: str = "registry.json") -> dict:
    """
    生成连接器注册表，支持增量更新
    
    Args:
        connectors_dir: connectors目录路径
        output_file: 输出文件路径
    
    Returns:
        生成的注册表字典
    """
    # 切换到connectors目录
    original_dir = os.getcwd()
    os.chdir(connectors_dir)
    
    try:
        # 扫描所有连接器配置文件（official和community）
        config_files = glob.glob('official/*/connector.json') + glob.glob('community/*/connector.json')
        print(f"Found {len(config_files)} connector configs: {config_files}")
        
        # 读取现有注册表（如果存在）
        existing_registry = {}
        if os.path.exists(output_file):
            try:
                with open(output_file, 'r') as f:
                    existing_registry = json.load(f)
                print(f"📖 Loaded existing registry with {len(existing_registry.get('connectors', {}))} connectors")
            except Exception as e:
                print(f"⚠️ Failed to load existing registry: {e}, creating new one")
        
        # 初始化新注册表结构
        registry = {
            'schema_version': '1.0',
            'last_updated': datetime.now(timezone.utc).isoformat(),
            'connectors': existing_registry.get('connectors', {}),
            'metadata': {
                'repository': os.environ.get('GITHUB_REPOSITORY', 'laofahai/linch-mind'),
                'commit': os.environ.get('GITHUB_SHA', 'unknown'),
                'total_count': 0
            }
        }
        
        # 处理每个连接器配置
        for config_path in config_files:
            try:
                with open(config_path, 'r') as f:
                    config = json.load(f)
                
                connector_id = config['id']
                connector_type = config_path.split('/')[0]  # official or community
                current_version = config['version']
                
                # 检查连接器是否已存在
                existing_connector = registry['connectors'].get(connector_id)
                
                if existing_connector:
                    existing_version = existing_connector.get('version', '0.0.0')
                    if existing_version != current_version:
                        print(f"🔄 Updating {connector_id}: {existing_version} -> {current_version}")
                        action = "updated"
                    else:
                        print(f"✅ Keeping {connector_id} v{current_version} (no changes)")
                        action = "kept"
                else:
                    print(f"🆕 Adding new connector: {connector_id} v{current_version}")
                    action = "added"
                
                # 构建连接器信息（新增或更新）
                connector_info = {
                    'id': config['id'],
                    'name': config['name'],
                    'version': config['version'],
                    'description': config['description'],
                    'author': config['author'],
                    'category': config['category'],
                    'type': connector_type,
                    'platforms': config.get('platforms', {}),
                    'permissions': config.get('permissions', []),
                    'capabilities': config.get('capabilities', {}),
                    'config_path': config_path,
                    'last_updated': datetime.now(timezone.utc).isoformat(),
                    'action': action
                }
                
                # 保留现有的下载URL（如果存在）
                if existing_connector and 'download_url' in existing_connector:
                    connector_info['download_url'] = existing_connector['download_url']
                
                registry['connectors'][connector_id] = connector_info
                
            except Exception as e:
                print(f"❌ Error processing {config_path}: {e}")
                continue
        
        registry['metadata']['total_count'] = len(registry['connectors'])
        
        # 保存注册表
        with open(output_file, 'w') as f:
            json.dump(registry, f, indent=2, ensure_ascii=False)
        
        print(f'✅ Registry generated with {len(registry["connectors"])} connectors')
        print(f'📄 Registry saved to: {output_file}')
        
        return registry
        
    finally:
        os.chdir(original_dir)


def update_download_urls(registry_file: str, release_tag: str, base_url: str = None) -> bool:
    """
    更新注册表中的下载URL
    
    Args:
        registry_file: 注册表文件路径
        release_tag: GitHub Release标签
        base_url: 基础URL，默认从环境变量获取
    
    Returns:
        更新是否成功
    """
    if not os.path.exists(registry_file):
        print(f"❌ Registry file not found: {registry_file}")
        return False
    
    # 构建基础URL
    if not base_url:
        repo = os.environ.get('GITHUB_REPOSITORY', 'laofahai/linch-mind')
        base_url = f"https://github.com/{repo}/releases/download/{release_tag}"
    
    try:
        with open(registry_file, 'r') as f:
            registry = json.load(f)
        
        updated_count = 0
        for connector_id, connector_info in registry['connectors'].items():
            zip_filename = f"{connector_id}-connector.zip"
            download_url = f"{base_url}/{zip_filename}"
            
            # 更新或添加下载URL
            old_url = connector_info.get('download_url')
            connector_info['download_url'] = download_url
            
            if old_url != download_url:
                print(f"🔗 Updated download URL for {connector_id}")
                updated_count += 1
        
        # 更新注册表时间戳
        registry['last_updated'] = datetime.now(timezone.utc).isoformat()
        registry['metadata']['release_tag'] = release_tag
        
        # 保存更新的注册表
        with open(registry_file, 'w') as f:
            json.dump(registry, f, indent=2, ensure_ascii=False)
        
        print(f"✅ Updated {updated_count} download URLs in registry")
        return True
        
    except Exception as e:
        print(f"❌ Error updating download URLs: {e}")
        return False


def calculate_checksum(file_path: str) -> str:
    """
    计算文件SHA256校验和
    """
    sha256_hash = hashlib.sha256()
    try:
        with open(file_path, "rb") as f:
            for byte_block in iter(lambda: f.read(4096), b""):
                sha256_hash.update(byte_block)
        return sha256_hash.hexdigest()
    except Exception as e:
        print(f"⚠️ Failed to calculate checksum for {file_path}: {e}")
        return ""


def main():
    parser = argparse.ArgumentParser(description='连接器注册表生成工具')
    parser.add_argument('--dir', default='.', help='connectors目录路径')
    parser.add_argument('--output', default='registry.json', help='输出文件名')
    parser.add_argument('--format', action='store_true', help='格式化输出')
    parser.add_argument('--update-urls', help='更新下载URL，需要提供release标签')
    parser.add_argument('--base-url', help='自定义基础URL')
    
    args = parser.parse_args()
    
    # 更新下载URL模式
    if args.update_urls:
        if not Path(args.output).exists():
            print(f"❌ 注册表文件不存在: {args.output}")
            sys.exit(1)
        
        success = update_download_urls(args.output, args.update_urls, args.base_url)
        sys.exit(0 if success else 1)
    
    # 正常生成模式
    if not Path(args.dir).exists():
        print(f"❌ 目录不存在: {args.dir}")
        sys.exit(1)
    
    registry = generate_registry(args.dir, args.output)
    
    if args.format:
        print("\n📋 Registry Summary:")
        print(f"   Total connectors: {registry['metadata']['total_count']}")
        for connector_id, info in registry['connectors'].items():
            action_emoji = {"added": "🆕", "updated": "🔄", "kept": "✅"}.get(info.get('action', 'kept'), '🔧')
            download_status = "📥" if 'download_url' in info else "❓"
            print(f"   {action_emoji} {connector_id} v{info['version']} ({info['type']}) {download_status}")
            if 'download_url' in info:
                print(f"      📍 {info['download_url']}")


if __name__ == "__main__":
    main()

File: connectors/scripts/registry_generator.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <filesystem>
#include <chrono>
#include <iomanip>
#include <sstream>
#include <getopt.h>
#include <nlohmann/json.hpp>

using json = nlohmann::json;
namespace fs = std::filesystem;

std::string getCurrentTimestamp() {
    auto now = std::chrono::system_clock::now();
    auto time_t = std::chrono::system_clock::to_time_t(now);
    std::stringstream ss;
    ss << std::put_time(std::gmtime(&time_t), "%Y-%m-%dT%H:%M:%SZ");
    return ss.str();
}

std::string getEnvVar(const std::string& name, const std::string& defaultValue = "") {
    const char* value = std::getenv(name.c_str());
    return value ? std::string(value) : defaultValue;
}

json generateRegistry(const std::string& connectorsDir, const std::string& outputFile) {
    fs::current_path(connectorsDir);
    
    // Find all connector.json files
    std::vector<std::string> configFiles;
    
    if (fs::exists("official")) {
        for (const auto& entry : fs::recursive_directory_iterator("official")) {
            if (entry.path().filename() == "connector.json") {
                configFiles.push_back(entry.path().string());
            }
        }
    }
    
    if (fs::exists("community")) {
        for (const auto& entry : fs::recursive_directory_iterator("community")) {
            if (entry.path().filename() == "connector.json") {
                configFiles.push_back(entry.path().string());
            }
        }
    }
    
    std::cout << "Found " << configFiles.size() << " connector configs" << std::endl;
    
    // Read existing registry if it exists
    json existingRegistry;
    if (fs::exists(outputFile)) {
        try {
            std::ifstream file(outputFile);
            file >> existingRegistry;
            std::cout << "📖 Loaded existing registry with " 
                      << existingRegistry.value("connectors", json::object()).size() 
                      << " connectors" << std::endl;
        } catch (const std::exception& e) {
            std::cout << "⚠️ Failed to load existing registry: " << e.what() 
                      << ", creating new one" << std::endl;
        }
    }
    
    // Initialize new registry structure
    json registry;
    registry["schema_version"] = "1.0";
    registry["last_updated"] = getCurrentTimestamp();
    registry["connectors"] = existingRegistry.value("connectors", json::object());
    registry["metadata"] = {
        {"repository", getEnvVar("GITHUB_REPOSITORY", "laofahai/linch-mind")},
        {"commit", getEnvVar("GITHUB_SHA", "unknown")},
        {"total_count", 0}
    };
    
    // Process each connector config
    for (const auto& configPath : configFiles) {
        try {
            std::ifstream file(configPath);
            json config;
            file >> config;
            
            std::string connectorId = config["id"];
            std::string connectorType = configPath.substr(0, configPath.find('/'));
            std::string currentVersion = config["version"];
            
            // Check if connector already exists
            json existingConnector = registry["connectors"].value(connectorId, json::object());
            std::string action;
            
            if (!existingConnector.empty()) {
                std::string existingVersion = existingConnector.value("version", "0.0.0");
                if (existingVersion != currentVersion) {
                    std::cout << "🔄 Updating " << connectorId << ": " 
                              << existingVersion << " -> " << currentVersion << std::endl;
                    action = "updated";
                } else {
                    std::cout << "✅ Keeping " << connectorId << " v" << currentVersion 
                              << " (no changes)" << std::endl;
                    action = "kept";
                }
            } else {
                std::cout << "🆕 Adding new connector: " << connectorId 
                          << " v" << currentVersion << std::endl;
                action = "added";
            }
            
            // Build connector info
            json connectorInfo = {
                {"id", config["id"]},
                {"name", config["name"]},
                {"version", config["version"]},
                {"description", config["description"]},
                {"author", config["author"]},
                {"category", config["category"]},
                {"type", connectorType},
                {"platforms", config.value("platforms", json::object())},
                {"permissions", config.value("permissions", json::array())},
                {"capabilities", config.value("capabilities", json::object())},
                {"config_path", configPath},
                {"last_updated", getCurrentTimestamp()},
                {"action", action}
            };
            
            // Preserve existing download URL if exists
            if (!existingConnector.empty() && existingConnector.contains("download_url")) {
                connectorInfo["download_url"] = existingConnector["download_url"];
            }
            
            registry["connectors"][connectorId] = connectorInfo;
            
        } catch (const std::exception& e) {
            std::cout << "❌ Error processing " << configPath << ": " << e.what() << std::endl;
            continue;
        }
    }
    
    registry["metadata"]["total_count"] = registry["connectors"].size();
    
    // Save registry
    std::ofstream outFile(outputFile);
    outFile << registry.dump(2) << std::endl;
    
    std::cout << "✅ Registry generated with " << registry["connectors"].size() 
              << " connectors" << std::endl;
    std::cout << "📄 Registry saved to: " << outputFile << std::endl;
    
    return registry;
}

bool updateDownloadUrls(const std::string& registryFile, const std::string& releaseTag, 
                       const std::string& baseUrl = "") {
    if (!fs::exists(registryFile)) {
        std::cout << "❌ Registry file not found: " << registryFile << std::endl;
        return false;
    }
    
    std::string finalBaseUrl = baseUrl;
    if (finalBaseUrl.empty()) {
        std::string repo = getEnvVar("GITHUB_REPOSITORY", "laofahai/linch-mind");
        finalBaseUrl = "https://github.com/" + repo + "/releases/download/" + releaseTag;
    }
    
    try {
        std::ifstream file(registryFile);
        json registry;
        file >> registry;
        file.close();
        
        int updatedCount = 0;
        for (auto& [connectorId, connectorInfo] : registry["connectors"].items()) {
            // Support multi-platform downloads
            json platforms = json::object();
            
            // Check for platform-specific packages
            std::vector<std::string> supportedPlatforms = {"linux-x64", "macos-x64", "windows-x64"};
            
            for (const std::string& platform : supportedPlatforms) {
                std::string zipFilename = connectorId + "-connector-" + platform + ".zip";
                std::string downloadUrl = finalBaseUrl + "/" + zipFilename;
                
                platforms[platform] = {
                    {"download_url", downloadUrl},
                    {"supported", true},
                    {"last_updated", getCurrentTimestamp()}
                };
            }
            
            // Update connector info with platform support
            json oldPlatforms = connectorInfo.value("platforms", json::object());
            connectorInfo["platforms"] = platforms;
            
            // Keep backward compatibility with single download_url
            std::string defaultZip = connectorId + "-connector-linux-x64.zip";
            std::string defaultUrl = finalBaseUrl + "/" + defaultZip;
            connectorInfo["download_url"] = defaultUrl;
            
            if (oldPlatforms != platforms) {
                std::cout << "🔗 Updated platform URLs for " << connectorId << std::endl;
                updatedCount++;
            }
        }
        
        registry["last_updated"] = getCurrentTimestamp();
        registry["metadata"]["release_tag"] = releaseTag;
        
        std::ofstream outFile(registryFile);
        outFile << registry.dump(2) << std::endl;
        
        std::cout << "✅ Updated " << updatedCount << " download URLs in registry" << std::endl;
        return true;
        
    } catch (const std::exception& e) {
        std::cout << "❌ Error updating download URLs: " << e.what() << std::endl;
        return false;
    }
}

void printUsage(const char* programName) {
    std::cout << "Usage: " << programName << " [options]" << std::endl;
    std::cout << "Options:" << std::endl;
    std::cout << "  --dir <path>        Connectors directory path (default: .)" << std::endl;
    std::cout << "  --output <file>     Output file name (default: registry.json)" << std::endl;
    std::cout << "  --format            Format output with summary" << std::endl;
    std::cout << "  --update-urls <tag> Update download URLs with release tag" << std::endl;
    std::cout << "  --base-url <url>    Custom base URL for downloads" << std::endl;
    std::cout << "  --help              Show this help message" << std::endl;
}

int main(int argc, char* argv[]) {
    std::string connectorsDir = ".";
    std::string outputFile = "registry.json";
    bool formatOutput = false;
    std::string updateUrls;
    std::string baseUrl;
    
    static struct option longOptions[] = {
        {"dir", required_argument, 0, 'd'},
        {"output", required_argument, 0, 'o'},
        {"format", no_argument, 0, 'f'},
        {"update-urls", required_argument, 0, 'u'},
        {"base-url", required_argument, 0, 'b'},
        {"help", no_argument, 0, 'h'},
        {0, 0, 0, 0}
    };
    
    int optionIndex = 0;
    int c;
    
    while ((c = getopt_long(argc, argv, "d:o:fu:b:h", longOptions, &optionIndex)) != -1) {
        switch (c) {
            case 'd':
                connectorsDir = optarg;
                break;
            case 'o':
                outputFile = optarg;
                break;
            case 'f':
                formatOutput = true;
                break;
            case 'u':
                updateUrls = optarg;
                break;
            case 'b':
                baseUrl = optarg;
                break;
            case 'h':
                printUsage(argv[0]);
                return 0;
            default:
                printUsage(argv[0]);
                return 1;
        }
    }
    
    try {
        // Update URLs mode
        if (!updateUrls.empty()) {
            if (!fs::exists(outputFile)) {
                std::cout << "❌ Registry file does not exist: " << outputFile << std::endl;
                return 1;
            }
            
            bool success = updateDownloadUrls(outputFile, updateUrls, baseUrl);
            return success ? 0 : 1;
        }
        
        // Normal generation mode
        if (!fs::exists(connectorsDir)) {
            std::cout << "❌ Directory does not exist: " << connectorsDir << std::endl;
            return 1;
        }
        
        json registry = generateRegistry(connectorsDir, outputFile);
        
        if (formatOutput) {
            std::cout << "\n📋 Registry Summary:" << std::endl;
            std::cout << "   Total connectors: " << registry["metadata"]["total_count"] << std::endl;
            
            for (const auto& [connectorId, info] : registry["connectors"].items()) {
                std::string action = info.value("action", "kept");
                std::string actionEmoji = "🔧";
                if (action == "added") actionEmoji = "🆕";
                else if (action == "updated") actionEmoji = "🔄";
                else if (action == "kept") actionEmoji = "✅";
                
                std::string downloadStatus = info.contains("download_url") ? "📥" : "❓";
                
                std::cout << "   " << actionEmoji << " " << connectorId 
                          << " v" << info["version"] << " (" << info["type"] << ") " 
                          << downloadStatus << std::endl;
                
                if (info.contains("download_url")) {
                    std::cout << "      📍 " << info["download_url"] << std::endl;
                }
            }
        }
        
        return 0;
        
    } catch (const std::exception& e) {
        std::cout << "❌ Error: " << e.what() << std::endl;
        return 1;
    }
}